File: DsWorkNet/Dswork.Core/Page/Page.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Dswork.Core.Page
{
	/// <summary>
	/// 分页信息, 第一页从1开始
	/// </summary>
	/// <typeparam name="T">IList的每条记录使用HashTable存储</typeparam>
	public class Page<T>
	{
		private IList resultList = (IList)new List<T>();
		private IList<T> result = new List<T>();
		private int pageSize;
		private int currentPage;
		private int lastPage;
		private int totalCount = 0;
		private String pageName = "page";
		private String pageSizeName = "pageSize";

		private void PageSetting(int currentPage, int pageSize, int totalCount)
		{
			if(pageSize <= 0)
			{
				throw new Exception("[pageSize] must great than zero");
			}
			this.lastPage = totalCount % pageSize == 0 ? totalCount / pageSize : totalCount / pageSize + 1;
			this.lastPage = (this.lastPage < 1) ? 1 : this.lastPage;
			this.pageSize = pageSize;
			if(currentPage <= 1)
			{
				currentPage = 1;
			}
			else if(int.MaxValue == currentPage || currentPage > this.lastPage)
			{
				currentPage = this.lastPage;
			}
			//		// 下面这种的话，如果pageSize=0的话，就相当于不分页
			//		if(pageSize == 0)
			//		{
			//			pageNumber = 1;
			//			pageSize = (totalCount <= 0) ? 1 : totalCount;
			//		}
			this.currentPage = currentPage;
			this.totalCount = totalCount;
		}

		/// <summary>
		/// 构造函数
		/// </summary>
		/// <param name="currentPage">当前页码</param>
		/// <param name="pageSize">一页显示的条数</param>
		/// <param name="totalCount">数据总条数</param>
		public Page(int currentPage, int pageSize, int totalCount)
		{
			PageSetting(currentPage, pageSize, totalCount);
		}

		/// <summary>
		/// 构造函数，使用ResultList变量，尽量使用泛型
		/// </summary>
		/// <param name="currentPage">当前页码</param>
		/// <param name="pageSize">一页显示的条数</param>
		/// <param name="totalCount">数据总条数</param>
		/// <param name="result">结果集IList</param>
		[Obsolete("Recommended to use Page(int currentPage, int pageSize, int totalCount, IList<T> result)", false)]
		public Page(int currentPage, int pageSize, int totalCount, IList result)
		{
			PageSetting(currentPage, pageSize, totalCount);
			this.SetResult(result);
		}

		/// <summary>
		/// 构造函数，使用Result变量&lt;T&gt;
		/// </summary>
		/// <param name="currentPage">当前页码</param>
		/// <param name="pageSize">一页显示的条数</param>
		/// <param name="totalCount">数据总条数</param>
		/// <param name="iresult">结果集IList<T></param>
		public Page(int currentPage, int pageSize, int totalCount, IList<T> result)
		{
			PageSetting(currentPage, pageSize, totalCount);
			this.SetResult(result);
			this.SetResult((IList)result);
		}

		/// <summary>
		/// 取得当前页页码
		/// </summary>
		public int CurrentPage
		{
			get
			{
				return currentPage;
			}
		}

		/// <summary>
		/// 得到结果集第一条数据的行编码
		/// </summary>
		public int FirstResultIndex
		{
			get
			{
				if(pageSize <= 0)
				{
					throw new Exception("[pageSize] must great than zero");
				}
				return (currentPage - 1) * pageSize;
			}
		}

		/// <summary>
		/// 取得数据总页数，也就是最后一页页码
		/// </summary>
		public int LastPage
		{
			get
			{
				return lastPage;
			}
		}

		/// <summary>
		/// 取得下一页页码
		/// </summary>
		public int NextPage
		{
			get
			{
				return currentPage + 1;
			}
		}

		/// <summary>
		/// 取得当前页码的参数名，默认值是page
		/// </summary>
		public String PageName
		{
			get
			{
				return pageName;
			}
			set
			{
				if(value == null || value.Trim().Length == 0)
				{
					throw new Exception("[pageName] must be not null");
				}
				pageName = value.Trim();
			}
		}

		/// <summary>
		/// 取得一页显示的条数
		/// </summary>
		public int PageSize
		{
			get
			{
				return pageSize;
			}
		}

		/// <summary>
		/// 取得一页显示的条数的参数名，默认值是page
		/// </summary>
		public String PageSizeName
		{
			get
			{
				return pageSizeName;
			}
			set
			{
				if(value == null || value.Trim().Length == 0)
				{
					throw new Exception("[pageSizeName] must be not null");
				}
				pageSizeName = value.Trim();
			}
		}

		/// <summary>
		/// 取得上一页页码
		/// </summary>
		public int PreviousPage
		{
			get
			{
				return currentPage - 1;
			}
		}

		/// <summary>
		/// 设置结果集
		/// </summary>
		/// <param name="value">IList</param>
		public void SetResult(IList value)
		{
			if(value == null)
			{
				throw new Exception("'result' must be not null");
			}
			resultList = value;
		}

		/// <summary>
		/// 设置结果集
		/// </summary>
		/// <param name="value">IList&lt;T&gt;</param>
		public void SetResult(IList<T> value)
		{
			if(value == null)
			{
				throw new Exception("'result<T>' must be not null");
			}
			result = value;
		}

		/// <summary>
		/// 取得结果集，结果集IList
		/// </summary>
		public IList GetResult()
		{
			return resultList;
		}

		/// <summary>
		/// 取得结果集，结果集IList&lt;T&gt;
		/// </summary>
		public IList<T> GetResult<M>()
		{
			return result;
		}

		/// <summary>
		/// 取得数据总条数，0表示没有数据
		/// </summary>
		public int TotalCount
		{
			get
			{
				return totalCount;
			}
		}

		/// <summary>
		/// 取得数据总页数
		/// </summary>
		public int TotalPage
		{
			get
			{
				int page = totalCount / pageSize;
				int tmp = totalCount % pageSize;
				return page + ((tmp == 0) ? 0 : 1);
			}
		}

		/// <summary>
		/// 是否是首页（第一页），第一页页码为1
		/// </summary>
		public Boolean IsFirstPage
		{
			get
			{
				return currentPage == 1;
			}
		}

		/// <summary>
		/// 是否有下一页
		/// </summary>
		public Boolean IsHasNextPage
		{
			get
			{
				return lastPage > currentPage;
			}
		}

		/// <summary>
		/// 是否有上一页
		/// </summary>
		public Boolean IsHasPreviousPage
		{
			get
			{
				return currentPage > 1;
			}
		}

		/// <summary>
		/// 是否是最后一页
		/// </summary>
		public Boolean IsLastPage
		{
			get
			{
				return currentPage >= lastPage;
			}
		}

	}
}

File: DsWorkJava/dswork.core/dswork/core/page/Page.java
package dswork.core.page;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * 分页信息, 第一页从1开始
 * @author skey
 */
public class Page<T> implements Serializable
{
	private static final long serialVersionUID = 1L;
	private List<T> result = new ArrayList<T>();
	private int pageSize;
	private int currentPage;
	private int lastPage;
	private int totalCount = 0;
	private String pageName = "page";
	private String pageSizeName = "pageSize";

	/**
	 * 构造函数
	 * @param currentPage 当前页码
	 * @param pageSize 一页显示的条数
	 * @param totalCount 数据总条数
	 */
	public Page(int currentPage, int pageSize, int totalCount)
	{
		this(currentPage, pageSize, totalCount, new ArrayList<T>(0));
	}

	/**
	 * 构造函数
	 * @param currentPage 当前页码
	 * @param pageSize 一页显示的条数
	 * @param totalCount 数据总条数
	 * @param result 结果集List&lt;T&gt;，其长度不影响计数
	 */
	public Page(int currentPage, int pageSize, int totalCount, List<T> result)
	{
		if(pageSize <= 0)
		{
			throw new IllegalArgumentException("[pageSize] must great than zero");
		}
		this.lastPage = totalCount % pageSize == 0 ? totalCount / pageSize : totalCount / pageSize + 1;
		if(this.lastPage < 1)
		{
			this.lastPage = 1;
		}
		this.pageSize = pageSize;
		if(currentPage <= 1)
		{
			currentPage = 1;
		}
		else if(Integer.MAX_VALUE == currentPage || currentPage > this.lastPage)
		{
			currentPage = this.lastPage;
		}
//		// 下面这种的话，如果pageSize=0的话，就相当于不分页
//		if(pageSize == 0)
//		{
//			pageNumber = 1;
//			pageSize = (totalCount <= 0) ? 1 : totalCount;
//		}
		this.currentPage = currentPage;
		this.totalCount = totalCount;
		setResult(result);
	}

	/**
	 * 取得当前页页码
	 * @return int
	 */
	public int getCurrentPage()
	{
		return currentPage;
	}

	/**
	 * 得到结果集第一条数据的行编码
	 * @return int
	 */
	public int getFirstResultIndex()
	{
		if (pageSize <= 0)
		{
			throw new IllegalArgumentException("[pageSize] must great than zero");
		}
		return (currentPage - 1) * pageSize;
	}
	
	/**
	 * 取得数据总页数，也就是最后一页页码
	 * @return int
	 */
	public int getLastPage()
	{
		return lastPage;
	}

	/**
	 * 取得下一页页码
	 * @return int
	 */
	public int getNextPage()
	{
		return currentPage + 1;
	}

	/**
	 * 取得当前页码的参数名
	 * @return String，默认值是page
	 */
	public String getPageName()
	{
		return pageName;
	}

	/**
	 * 设置当前页码的参数名，默认值是page
	 * @param pageName String
	 */
	public void setPageName(String pageName)
	{
		if (pageName == null || pageName.trim().length() == 0)
		{
			throw new IllegalArgumentException("[pageName] must be not null");
		}
		this.pageName = pageName.trim().replace("\"", "&quot;").replace("\r", "").replace("\n", "");
	}

	/**
	 * 取得一页显示的条数
	 * @return int
	 */
	public int getPageSize()
	{
		return pageSize;
	}

	/**
	 * 取得一页显示的条数的参数名
	 * @return String，默认值是pageSize
	 */
	public String getPageSizeName()
	{
		return pageSizeName;
	}

	/**
	 * 设置当前页码的参数名，默认值是pageSize
	 * @param pageSizeName String
	 */
	public void setPageSizeName(String pageSizeName)
	{
		if (pageSizeName == null || pageSizeName.trim().length() == 0)
		{
			throw new IllegalArgumentException("[pageSizeName] must be not null");
		}
		this.pageSizeName = pageSizeName.trim().replace("\"", "&quot;").replace("\r", "").replace("\n", "");
	}

	/**
	 * 取得上一页页码
	 * @return int
	 */
	public int getPreviousPage()
	{
		return currentPage - 1;
	}

	/**
	 * 取得结果集
	 * @return 结果集List&lt;T&gt;
	 */
	public List<T> getResult()
	{
		return result;
	}

	/**
	 * 设置结果集，其值与长度不影响计数
	 * @param result 结果集List&lt;T&gt;
	 */
	public void setResult(List<T> result)
	{
		if(result == null)
		{
			throw new IllegalArgumentException("'result' must be not null");
		}
		this.result = result;
	}

	/**
	 * 取得数据总条数，0表示没有数据
	 * @return int
	 */
	public int getTotalCount()
	{
		return totalCount;
	}

	/**
	 * 取得数据总页数
	 * @return int
	 */
	public int getTotalPage()
	{
		return lastPage;
	}

	/**
	 * 是否是首页（第一页），第一页页码为1
	 * @return boolean
	 */
	public boolean isFirstPage()
	{
		return currentPage == 1;
	}

	/**
	 * 是否有下一页
	 * @return boolean
	 */
	public boolean isHasNextPage()
	{
		return lastPage > currentPage;
	}

	/**
	 * 是否有上一页
	 * @return boolean
	 */
	public boolean isHasPreviousPage()
	{
		return currentPage > 1;
	}

	/**
	 * 是否是最后一页
	 * @return boolean
	 */
	public boolean isLastPage()
	{
		return currentPage >= lastPage;
	}

//	/**
//	 * 取得当前页的首条数据的行编码
//	 * @return int
//	 */
//	public int getThisPageFirstElementNumber()
//	{
//		return (currentPage - 1) * pageSize + 1;
//	}
//
//	/**
//	 * 取得当前页的末条数据的行编码
//	 * @return int
//	 */
//	public int getThisPageLastElementNumber()
//	{
//		int fullPage = getThisPageFirstElementNumber() + pageSize - 1;
//		return totalCount < fullPage ? totalCount : fullPage;
//	}

//	/**
//	 * 得到用于多页跳转的页码
//	 * @return
//	 */
//	public List<Integer> getLinkPageNumbers()
//	{
//		return generateLinkPageNumbers(getCurrentPage(), getLastPage(), 10);
//	}

//	private static List<Integer> generateLinkPageNumbers(int currentPageNumber, int lastPageNumber, int count)
//	{
//		int avg = count / 2;
//		int startPageNumber = currentPageNumber - avg;
//		if (startPageNumber <= 0)
//		{
//			startPageNumber = 1;
//		}
//		int endPageNumber = startPageNumber + count - 1;
//		if (endPageNumber > lastPageNumber)
//		{
//			endPageNumber = lastPageNumber;
//		}
//		if (endPageNumber - startPageNumber < count)
//		{
//			startPageNumber = endPageNumber - count;
//			if (startPageNumber <= 0)
//			{
//				startPageNumber = 1;
//			}
//		}
//		List<Integer> result = new ArrayList<Integer>();
//		for (int i = startPageNumber; i <= endPageNumber; i++)
//		{
//			result.add(new Integer(i));
//		}
//		return result;
//	}
}
